src/server/search_family.cc

```diff
@@ -0,0 +1,244 @@
+using SerializedDocument = pair<std::string /*key*/, DocumentData>;
+using Query = search::AstExpr;
+
+struct BaseAccessor : public search::HSetAccessor {
+  using FieldConsumer = search::HSetAccessor::FieldConsumer;
+
+  virtual DocumentData Serialize() const = 0;
+};
+
+struct ListPackAccessor : public BaseAccessor {
+  using LpPtr = uint8_t*;
+
+  ListPackAccessor(LpPtr ptr) : lp_{ptr} {
+  }
+
+  bool Check(FieldConsumer f, string_view active_field) const override {
+    uint8_t intbuf[LP_INTBUF_SIZE];
+
+    if (!active_field.empty()) {
+      return f(container_utils::LpFind(lp_, active_field, intbuf).value_or(""));
+    }
+
+    uint8_t* fptr = lpFirst(lp_);
+    DCHECK_NE(fptr, nullptr);
+
+    while (fptr) {
+      fptr = lpNext(lp_, fptr);  // skip key
+      string_view v = container_utils::LpGetView(fptr, intbuf);
+      fptr = lpNext(lp_, fptr);
+
+      if (f(v))
+        return true;
+    }
+
+    return false;
+  }
+
+  DocumentData Serialize() const override {
+    uint8_t intbuf[2][LP_INTBUF_SIZE];
+    DocumentData out{};
+
+    uint8_t* fptr = lpFirst(lp_);
+    DCHECK_NE(fptr, nullptr);
+
+    while (fptr) {
+      string_view k = container_utils::LpGetView(fptr, intbuf[0]);
+      fptr = lpNext(lp_, fptr);  // skip key
+      string_view v = container_utils::LpGetView(fptr, intbuf[1]);
+      fptr = lpNext(lp_, fptr);
+
+      out[k] = v;
+    }
+
+    return out;
+  }
+
+ private:
+  LpPtr lp_;
+};
+
+struct StringMapAccessor : public BaseAccessor {
+  StringMapAccessor(StringMap* hset) : hset_{hset} {
+  }
+
+  bool Check(FieldConsumer f, string_view active_field) const override {
+    if (!active_field.empty()) {
+      return f(SdsToSafeSv(hset_->Find(active_field)));
+    }
+
+    for (const auto& [k, v] : *hset_) {
+      if (f(SdsToSafeSv(v)))
+        return true;
+    }
+
+    return false;
+  }
+
+  DocumentData Serialize() const override {
+    DocumentData out{};
+    for (const auto& [k, v] : *hset_)
+      out[SdsToSafeSv(k)] = SdsToSafeSv(v);
+    return out;
+  }
+
+ private:
+  StringMap* hset_;
+};
+
+unique_ptr<BaseAccessor> GetAccessor(const OpArgs& op_args, const PrimeValue& pv) {
+  if (pv.Encoding() == kEncodingListPack) {
+    auto ptr = reinterpret_cast<ListPackAccessor::LpPtr>(pv.RObjPtr());
+    return make_unique<ListPackAccessor>(ptr);
+  } else {
+    DCHECK_EQ(pv.Encoding(), kEncodingStrMap2);
+    auto* sm = container_utils::GetStringMap(pv, op_args.db_cntx);
+    return make_unique<StringMapAccessor>(sm);
+  }
+}
+
+// Perform brute force search for all hashes in shard with specific prefix
+// that match the query
+void OpSearch(const OpArgs& op_args, string_view prefix, const Query& query,
+              vector<SerializedDocument>* shard_out) {
+  auto& db_slice = op_args.shard->db_slice();
+  DCHECK(db_slice.IsDbValid(op_args.db_cntx.db_index));
+  auto [prime_table, _] = db_slice.GetTables(op_args.db_cntx.db_index);
+
+  string scratch;
```

Comment:
Shouldn't `scratch` be inside the callback? I kinda think that there's a race here, and in any case I don't see why to extend its lifetime beyond it?

Reply:
table->Traverse is just plain cursor based iteration with a callback - no threads/fibers involved. Having scratch outside the callback allows re-using the internal buffer, so we don't need to allocate a new one for every key read.

Reply:
Gotcha!